raft/broadcast_tables/experimental/lang.cc

```diff
@@ -22,7 +24,95 @@ bool is_broadcast_table_statement(const sstring& keyspace, const sstring& column
+}
+
+future<> execute_broadcast_table_query(
+    service::storage_proxy& proxy,
+    const query& query,
+    utils::UUID cmd_id) {
+    return std::visit(make_visitor (
+    std::bind_front([] (service::storage_proxy& proxy, const raft::broadcast_tables::select_query& q) -> future<> {
+        const auto schema = db::system_keyspace::broadcast_kv_store();
+        const auto* column_definition = schema->get_column_definition("value");
+
+        // Read mutations
+        const auto [read_cmd, range] = prepare_read_command(proxy, *schema, q.key);
+        auto [rs, _] = co_await proxy.query_mutations_locally(schema, read_cmd, range, db::no_timeout);
+
+        if (rs->partitions().empty()) {
+            co_return;
+        }
+
+        assert(rs->partitions().size() == 1); // In this version only one value per partition key is allowed.
+
+        const auto& p = rs->partitions()[0];
+        auto mutation = p.mut().unfreeze(schema);
+        const auto cell = get_atomic_cell(schema, mutation, "value");
+
+        // TODO return result
+    }, std::ref(proxy)),
+    std::bind_front([] (service::storage_proxy& proxy, utils::UUID cmd_id, const raft::broadcast_tables::update_query& q) -> future<> {
+        const auto schema = db::system_keyspace::broadcast_kv_store();
+
+        // Read mutations
+        const auto [read_cmd, range] = prepare_read_command(proxy, *schema, q.key);
+        auto [rs, _] = co_await proxy.query_mutations_locally(schema, read_cmd, range, db::no_timeout);
+
+        bool found = !rs->partitions().empty();
+
+        assert(!found || rs->partitions().size() == 1); // In this version at most one value per partition key is allowed.
+
+        auto new_mutation = found
+            ? rs->partitions()[0].mut().unfreeze(schema)
+            : mutation(schema, partition_key::from_single_value(*schema, to_bytes(q.key)));
+
+        const auto cell = found
+            ? std::optional<atomic_cell_view>{get_atomic_cell(schema, new_mutation, "value")}
+            : std::nullopt;
+
+        const auto previous_value = found
+            ? bytes_opt{cell->value().linearize()}
+            : std::nullopt;
+
+        bool is_conditional = q.value_condition.has_value();
+        bool is_applied = !is_conditional || *q.value_condition == previous_value;
+
+        if (is_applied) {
+            auto old_ts = found ? cell->timestamp() : api::min_timestamp;
+            auto from_state_id = utils::UUID_gen::micros_timestamp(cmd_id);
+            auto ts = std::max(from_state_id, old_ts + 1);
+
+            auto value = utf8_type->deserialize(q.new_value);
+            new_mutation.set_clustered_cell(clustering_key::make_empty(), "value", std::move(value), ts);
+            co_await proxy.mutate_locally(new_mutation, {}, {});
+        }
+    }, std::ref(proxy), cmd_id)), query.q);
```

Comment:
I guess you're using bind_front because the outer function isn't a coroutine and the inner functions are coroutines, and you want to avoid losing the captures? Better make a comment about it.

Reply:
> I guess you're using bind_front because the outer function isn't a coroutine and the inner functions are coroutines, and you want to avoid losing the captures?

Yes, added a comment.